### PR TITLE
Add checks

### PR DESCRIPTION
**Pull request by jcugat**
I noticed while reviewing https://github.com/encode/httpcore/pull/153 that some imports were not ordered correctly and some of them were unused. I've pretty much copied the approach from `httpx` regarding the scripts and configs.

Depends on https://github.com/encode/httpcore/pull/153 since one of the check failures is regarding `max_keepalive_connections` which will be fixed in the other PR, and wanted to avoid any conflicts. Funny enough, if we had `flake8` enabled before we would have noticed that bug 😅

### CODE REVIEW

File: httpcore/__init__.py [hunk 1]
```diff
@@ -24,29 +24,31 @@
 from ._sync.http_proxy import SyncHTTPProxy
 
 __all__ = [
```
Comment (jcugat) [inline httpcore/__init__.py:26]: The list has been sorted alphabetically, and added `ProtocolError` and `ProxyError` that were missing I guess due to a mistake.

File: httpcore/_backends/sync.py [hunk 2]
```diff
@@ -137,10 +137,9 @@ def open_tcp_stream(
         exc_map = {socket.timeout: ConnectTimeout, socket.error: ConnectError}
 
         with map_exceptions(exc_map):
-            local_addrport = None
```
Comment (jcugat) [inline httpcore/_backends/sync.py:140]: `local_addrport` was unused, leftover?

File: requirements.txt [hunk 3]
```diff
@@ -5,23 +5,26 @@ trio
...
 mkautodoc
+mkdocs
 mkdocs-material
 
 # Packaging
 twine
 wheel
 
 # Tests & Linting
-pytest
```
Comment (jcugat) [inline requirements.txt:17]: List ordered, and added the `flake8*` packages.

File: scripts/check [hunk 4]
```diff
@@ -8,5 +8,8 @@ export SOURCE_FILES="httpcore tests"
 
 set -x
 
-${PREFIX}mypy --ignore-missing-imports httpcore
+${PREFIX}isort --check --diff --project=httpx $SOURCE_FILES
+${PREFIX}black --check --diff $SOURCE_FILES
+${PREFIX}flake8 $SOURCE_FILES
+${PREFIX}mypy $SOURCE_FILES
```
Comment (jcugat) [inline scripts/check:14]: `--ignore-missing-imports` argument removed since it's already in `setup.cfg`

File: scripts/test [hunk 5]
```diff
@@ -5,22 +5,12 @@ if [ -d 'venv' ] ; then
     export PREFIX="venv/bin/"
 fi
 
+set -ex
+
 if [ -z $GITHUB_ACTIONS ]; then
-    set +e
     scripts/check
-    while [ $? -ne 0 ]; do
-        read -p "Running 'scripts/check' failed. Do you want to run 'scripts/lint' now? [y/N]  " yn
```
Comment (jcugat) [inline scripts/test:12]: Removed this just like we did with `httpx` https://github.com/encode/httpx/pull/1107

File: scripts/lint [hunk 6]
```diff
@@ -10,7 +10,7 @@ set -x
 
 ${PREFIX}autoflake --in-place --recursive $SOURCE_FILES
 ${PREFIX}isort --project=httpcore $SOURCE_FILES
-${PREFIX}black --exclude 'httpcore/_sync/.*' --exclude 'tests/sync_tests/.*' --target-version=py36 $SOURCE_FILES
```
Comment (jcugat) [inline scripts/lint:13]: Moved arguments to `pyproject.toml`, this way is also picked up by IDEs.

File: httpcore/__init__.py (hunk 1 above)
Comment (florimondmanca) [reply httpcore/__init__.py:26]: > added ProtocolError and ProxyError that were missing I guess due to a mistake.

I guess we could add an extra validation step in `test_exceptions.py` for this, as a follow up? :-) Or maybe even a more general test case for "verify that all exported members of `httpcore` as identified by `vars(httpcore)` are present in `__all__`". Maybe even check that the list is sorted… (Without validation the list will go unsorted again soon!)

File: httpcore/_backends/sync.py (hunk 2 above)
Comment (florimondmanca) [reply httpcore/_backends/sync.py:140]: Yes, leftover from moving this code to computing `source_address` above. :+1:

File: tests/async_tests/test_interfaces.py [hunk 7]
```diff
@@ -313,7 +316,7 @@ async def test_connection_pool_get_connection_info(
     reason="Unix Domain Sockets only exist on Unix",
 )
 @pytest.mark.usefixtures("async_environment")
-async def test_http_request_unix_domain_socket(uds_server) -> None:
+async def test_http_request_unix_domain_socket(uds_server: Server,) -> None:
```
Comment (florimondmanca) [inline tests/async_tests/test_interfaces.py:319]: (Nit)

```suggestion
async def test_http_request_unix_domain_socket(uds_server: Server) -> None:
```

**Comment by florimondmanca** [comment]
> Funny enough, if we had flake8 enabled before we would have noticed that bug 😅

Indeed…

```console
+ flake8 httpcore tests
httpcore/_async/connection_pool.py:104:13: F841 local variable 'max_keepalive_connections' is assigned to but never used
##[error]Process completed with exit code 1.
```

Well done, flake8!

File: tests/async_tests/test_interfaces.py (hunk 7 above)
Comment (jcugat) [reply tests/async_tests/test_interfaces.py:319]: Weird, I thought black would normalize this.

File: httpcore/__init__.py (hunk 1 above)
Comment (jcugat) [reply httpcore/__init__.py:26]: That's a great idea, I've added the test for both the order and make sure that everything is in `__all__`.

**Comment by jcugat** [comment]
Completely forgot about #144 (and I even commented there!).

File: requirements.txt [hunk 8]
```diff
@@ -5,23 +5,26 @@ trio
...
-pytest-asyncio
-pytest-trio
-pytest-cov
-black
 autoflake
-mypy
+black
+flake8
+flake8-bugbear
+flake8-pie
```
Comment (JayH5) [inline requirements.txt:21]: I'm definitely +1 on flake8, but less keen on these extras, particularly `flake8-pie` is not very useful as 3/4 of its checks are for Celery. I know in the past @tomchristie has pushed back against using lots of extra linting tools.

Comment (jcugat) [reply requirements.txt:21]: I added those to be consistent with `httpx`

https://github.com/encode/httpx/blob/655773e1c1b75895eda927d5a9d22a3b5b8f572d/requirements.txt#L21

Comment (JayH5) [reply requirements.txt:21]: Ah I think I was misremembering when `flake8-comprehensions` got removed from httpx. To be clear, this is a minor nit. I'm happy with `flake8-bugbear`, but `flake8-pie` is literally one check ('PIE781: You are assigning to a variable and then returning. Instead remove the assignment and return.') which seems a bit unnecessary.

**Pull request merged**